console/v1/tests/consoleplugins.console.openshift.io/AAA_ungated.yaml

```diff
@@ -87,3 +87,176 @@ tests:
+      expectedError: "spec.contentSecurityPolicy[1]: Duplicate value: map[string]interface {}{\"directive\":\"ScriptSrc\""
+    - name: Should throw an error for duplicate CSP directive values
+      initial: |
+        apiVersion: console.openshift.io/v1
+        kind: ConsolePlugin
+        spec:
+          displayName: foo
+          backend:
+            type: Service
+          contentSecurityPolicy:
+          - directive: ScriptSrc
+            values:
+            - https://script1.com/
+            - https://script1.com/
+      expectedError: "spec.contentSecurityPolicy[0].values: Invalid value: \"array\": each CSP directive value must be unique"
+    - name: Should throw an error for incorrect directive type
+      initial: |
+        apiVersion: console.openshift.io/v1
+        kind: ConsolePlugin
+        spec:
+          displayName: foo
+          backend:
+            type: Service
+          contentSecurityPolicy:
+          - directive: TestSrc
+            values:
+            - https://script1.com/
+      expectedError: "spec.contentSecurityPolicy[0].directive: Unsupported value: \"TestSrc\": supported values: \"DefaultSrc\", \"ScriptSrc\", \"StyleSrc\", \"ImgSrc\", \"FontSrc\""
+    - name: Should throw an error for invalid CSP directive values, using wildcard
+      initial: |
+        apiVersion: console.openshift.io/v1
+        kind: ConsolePlugin
+        spec:
+          displayName: foo
+          backend:
+            type: Service
+          contentSecurityPolicy:
+          - directive: ScriptSrc
+            values:
+            - "*"
+      expectedError: "spec.contentSecurityPolicy[0].values[0]: Invalid value: \"string\": CSP directive value cannot be a wildcard"
+    - name: Should be able to create a ConsolePlugin with valid CSP value, using domain with wildcard prefix
+      initial: |
+        apiVersion: console.openshift.io/v1
+        kind: ConsolePlugin
+        spec:
+          displayName: foo
+          backend:
+            type: Service
+          contentSecurityPolicy:
+          - directive: ScriptSrc
+            values:
+            - "*.example.com"
+      expected: |
+        apiVersion: console.openshift.io/v1
+        kind: ConsolePlugin
+        spec:
+          displayName: foo
+          backend:
+            type: Service
+          contentSecurityPolicy:
+          - directive: ScriptSrc
+            values:
+            - "*.example.com"
+    - name: Should throw an error for invalid CSP directive values, starting and ending with quotes
+      initial: |
+        apiVersion: console.openshift.io/v1
+        kind: ConsolePlugin
+        spec:
+          displayName: foo
+          backend:
+            type: Service
+          contentSecurityPolicy:
+          - directive: ScriptSrc
+            values:
+            - "'none'"
+      expectedError: "spec.contentSecurityPolicy[0].values[0]: Invalid value: \"string\": CSP directive value cannot start or and with a quote"
```

Comment:
What about unbalanced quotes? What if I had
```
- a'bad
- thing'b
```

concatenated this becomes `a'bad thing'b`, is the `'bad thing'` in there problematic?

Reply:
If a CSP header like `Content-Security-Policy: script-src a'bad thing'b;` is used, the browser will consider it invalid due to the unescaped or unrecognized syntax in a'bad thing'b. The result is that the entire CSP directive could be ignored. That said I think we should disable the quote completely. If you agree will also keep the previous check for a value which starts and ends with quote?

Reply:
Do we allow hash values?

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/Sources#hash-algorithm-base64-value

We want to support that for plugins that have inline scripts. It requires using `'` characters, though. (nonce we shouldn't support, however, since there's no way to generate a unique value each time.)

Reply:
@spadgett then I think we need to define the validation rules as:
- Hash directive format: Check for a valid hash algorithm (like SHA-256) followed by a hyphen and a valid Base64 value.
- Host source format: Check that the input matches the pattern of a valid URL.

For the CEL rule below accomplishes this by checking for two conditions:

- The value starts with a valid hash algorithm (sha256, sha384, or sha512) followed by a - and a valid Base64 string.
- The value matches the pattern of a valid host source (beginning with http:// or https://).

WDYT ?
ccing @vojtechszocs

Reply:
Host sources don't need to start with `http://` or `https://`. For example, `*.example.com` is a valid host source.

Reply:
Right, sorry. I've unintentionally leaved it out, but it should be valid as well 👍

Reply:
@spadgett @jhadvig I think the currently proposed validation rules
```
// +kubebuilder:validation:MinLength=1
// +kubebuilder:validation:MaxLength=1024
// +kubebuilder:validation:XValidation:rule="!self.contains(\"'\")",message="CSP directive value cannot contain a quote"
// +kubebuilder:validation:XValidation:rule="!self.matches('\\\\s')",message="CSP directive value cannot contain a whitespace"
// +kubebuilder:validation:XValidation:rule="!self.contains(',')",message="CSP directive value cannot contain a comma"
// +kubebuilder:validation:XValidation:rule="!self.contains(';')",message="CSP directive value cannot contain a semi-colon"
// +kubebuilder:validation:XValidation:rule="self != '*'",message="CSP directive value cannot be a wildcard"
type CSPDirectiveValue string
```
are reasonably strict, we can always loosen these rules based on CSP violation reports once we start collecting them with telemetry.

Additional validation logic can be placed into the Console operator (since it's responsible for ensuring valid configuration of Console and all of its resources) or directly into Bridge [`server.go`](https://github.com/openshift/console/blob/master/pkg/server/server.go) function `indexHandler` where we generate the actual CSP string.

One example of additional validation would be e.g. `DefaultSrc` and `ScriptSrc` containing the same value, in which case we want to remove that value from `ScriptSrc` since the `DefaultSrc` directive acts as a fallback for the `ScriptSrc` directive.

Another example of additional validation would be to ensure the plugin provided CSP source values are valid URLs with potential `*` wildcards.

The only `'something'` exception that makes sense to me is [`'<hash-algorithm>-<base64-value>'`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/Sources#hash-algorithm-base64-value) but I think it's better to wait for CSP violation reports first before we start allowing such exceptions. (Other `'something'` special values don't seem too relevant to me at the moment, the main focus should be allowing additional plugin-specific remote sources so they can be loaded without issues in the browser.)

Reply:
I would agree that lets start with stricter validation and relax it down the road, since we will be in report mode for couple of releases.

Reply:
> That said I think we should disable the quote completely. If you agree will also keep the previous check for a value which starts and ends with quote?

If we are removing quotes completely, why would we need a separate check to check for starting and ending with quotes? 

If we need to allow the start and end quotes specifically just for the hash algorithm, we could relax the no quotes validation in a compatible way later I believe.

Reply:
@JoelSpeed sorry for the confusion, I've removed the the check for starting and ending with quotes as well in the end 👍

Reply:
I am good with disallowing single quotes to start and relaxing validation later if needed